cdap-api/src/main/java/co/cask/cdap/api/plugin/PluginClass.java

```diff
@@ -146,7 +146,10 @@ public String getConfigFieldName() {
    * @return the {@link Requirements} which represents the requirements of the plugin
    */
   public Requirements getRequirements() {
-    return requirements;
+    // CDAP-14515 Requirements can be null in the case when the cluster is upgraded and the the Plugin class
+    // is being constructed from a deserialized form and the GSON used to deserialize does not uses
+    // PluginClassDeserializer responsible for handling the absence of requirements in old deserialized form.
```

Comment:
Then why we have the `PluginClassDeserializer` class if not being used in everywhere that involves serde of `PluginClass`? It seems counter-intuitive to fix in here.

Reply:
I think it's safer to do this null check, as it's very easy to forget you need a custom deserializer, especially if it's a field of a field of the object you're actually deserializing. 

I think I made the PluginClassDeserializer to try and avoid these types of null checks and to validate that certain required fields are present. But the downsides outweigh the benefits, as seen from the fact that we have this bug to begin with. IMO it makes sense to remove this class and move the validation logic into PluginClass itself, with clients that handle user input performing validation, but everything internal trusting that it's a valid object.

Reply:
If removing the `PluginClassDeserializer` makes a better case, please remove that class and make the `PluginClass` robust to upgrade.